File: app/src/main/java/com/example/nikhil/bakingapp/widget/RecipeIngredientsWidgetProvider.java
package com.example.nikhil.bakingapp.widget;

import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProvider;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.widget.RemoteViews;

import com.example.nikhil.bakingapp.R;
import com.example.nikhil.bakingapp.networking.NetworkUtils;

/**
 * Created by NIKHIL on 24-12-2017.
 */

public class RecipeIngredientsWidgetProvider extends AppWidgetProvider {

    @Override
    public void onUpdate(Context context, AppWidgetManager appWidgetManager, int[] appWidgetIds) {
        for (int appWidgetId: appWidgetIds) {
            updateAppWidget(context, appWidgetManager, appWidgetId);
        }
    }

    @Override
    public void onEnabled(Context context) {
        // Enter relevant functionality for when the first widget is created
    }

    @Override
    public void onDisabled(Context context) {
        NetworkUtils.ingredientsSelected.clear();
    }

    public void updateAppWidget(Context context, AppWidgetManager appWidgetManager, int appWidgetId) {
        // Construct the RemoteViews object
        RemoteViews views = new RemoteViews(context.getPackageName(), R.layout.recipe_ingredients_widget);
        Intent intent = new Intent(context, IngredientsRemoteViewsService.class);
        views.setRemoteAdapter(R.id.widgetIngredientListView, intent);
        Bundle appWidgetOptions = appWidgetManager.getAppWidgetOptions(appWidgetId);
        String recipeName = "  ";
        if (appWidgetOptions.containsKey("recipeName")) {
            recipeName = appWidgetOptions.getString("recipeName");
        }
        views.setTextViewText(R.id.widgetRecipeName, recipeName);

        // Instruct the widget manager to update the widget
        appWidgetManager.updateAppWidget(appWidgetId, views);
    }
}

File: app/src/main/java/com/example/nikhil/bakingapp/widget/RecipeIngredientsWidgetProvider.kt
package com.example.nikhil.bakingapp.widget

import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProvider
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.widget.RemoteViews

import com.example.nikhil.bakingapp.R
import com.example.nikhil.bakingapp.networking.NetworkUtils

/**
 * Created by NIKHIL on 24-12-2017.
 */

class RecipeIngredientsWidgetProvider : AppWidgetProvider() {

    override fun onUpdate(context: Context, appWidgetManager: AppWidgetManager, appWidgetIds: IntArray) {
        for (appWidgetId in appWidgetIds) {
            updateAppWidget(context, appWidgetManager, appWidgetId)
        }
    }

    override fun onEnabled(context: Context) {
        // Enter relevant functionality for when the first widget is created
    }

    override fun onDisabled(context: Context) {
        NetworkUtils.ingredientsSelected.clear()
    }

    fun updateAppWidget(context: Context, appWidgetManager: AppWidgetManager, appWidgetId: Int) {
        // Construct the RemoteViews object
        val views = RemoteViews(context.packageName, R.layout.recipe_ingredients_widget)
        val intent = Intent(context, IngredientsRemoteViewsService::class.java)
        views.setRemoteAdapter(R.id.widgetIngredientListView, intent)
        val appWidgetOptions = appWidgetManager.getAppWidgetOptions(appWidgetId)
        var recipeName: String? = "  "
        if (appWidgetOptions.containsKey("recipeName")) {
            recipeName = appWidgetOptions.getString("recipeName")
        }
        views.setTextViewText(R.id.widgetRecipeName, recipeName)

        // Instruct the widget manager to update the widget
        appWidgetManager.updateAppWidget(appWidgetId, views)
    }
}
